omegaconf/listconfig.py

```diff
@@ -234,10 +244,16 @@ def append(self, item: Any) -> None:
             index = len(self)
             self._validate_set(key=index, value=item)
 
+            element_type = self.__dict__["_metadata"].element_type
+            value = item
+            expect_container = is_container_annotation(element_type)
+            if expect_container and isinstance(item, BaseContainer):
+                value = item._value()
+
             node = _maybe_wrap(
                 ref_type=self.__dict__["_metadata"].element_type,
                 key=index,
-                value=item,
+                value=value,
                 is_optional=OmegaConf.is_optional(item),
                 parent=self,
             )
```

Comment:
What is this doing?
It looks like if you get a DictConfig or ListConfig you use the content directly and wrap it again.
What is this achieving?

Reply:
As I said above(and other PRs), if you merge two containers you have to deal with maybe_wrap which won't wrap nodes. 

We want to type check whether the contents of the value if we expect the value to be a container, with certain element_type and key_type. The solution I found to work here and other parts of the code(like input_config above) is to unwrap to check again.

Reply:
A possibly cleaner solution is to treat the case of a DictConfig assigned onto a DictConfig by validating that the key type and the element types of the src are compatible with those in the target.

Maybe we can implement a function like:
```
is_legal_assignment(dest_type, src_type) -> bool
```

To handle that.
That function can be used to check if the type of a value can be assigned to a specific target type, but can also be used to check if `Dict[str, int]` can be assigned to `Dict[str, Any]` (yes) or to `Dict[int, int]` (no).

There is no need to re-wrap the DictConfig. we can know that all of keys and values are conforming to the specified key type and element type.

Reply:
Do you mean this function should check the type of every value in DictConfig?

Reply:
I mean that a DictConfig can be assumed to be legal.
There was validation performed on it's input when it was created.
When you are assigning a DictConfig, you should not need to recursively re-validate it.

You can look at the key type and element types:
if they are matching (or more specific), the assignment is legal.